packages/apps/plugins/plugin-space/src/SpacePlugin.tsx

```diff
@@ -1033,28 +1018,96 @@ export const SpacePlugin = ({
+                  object.objects.forEach((obj) => {
+                    if (!parentCollection.objects.includes(obj)) {
+                      parentCollection.objects.push(obj);
+                    }
+                  });
+                }
+
+                space.db.remove(object as any);
+
+                const undoMessageKey =
+                  object instanceof CollectionType ? 'collection deleted label' : 'object deleted label';
+
                 return {
-                  intents: [
-                    [
-                      {
-                        action: LayoutAction.SET_LAYOUT,
-                        data: {
-                          element: 'popover',
-                          anchorId: `dxos.org/ui/${caller}/${fullyQualifiedId(object)}`,
-                          component: 'dxos.org/plugin/space/RemoveObjectPopover',
-                          subject: {
-                            object,
-                            collection: intent.data?.collection,
-                          },
-                        },
-                      },
-                    ],
-                  ],
+                  data: true,
+                  undoable: {
+                    message: translations[0]['en-US'][SPACE_PLUGIN][undoMessageKey], // Consider using a translation key here
+                    data: deletionData,
+                  },
                 };
+              } else if (intent.undo) {
+                const undoData = intent.data;
+                if (undoData && undoData.object && undoData.parentCollection) {
+                  // Restore the object to the space
+                  const restoredObject = space.db.add(undoData.object as any);
+
+                  // Restore the object to its original position in the collection
+                  if (undoData.parentCollection instanceof CollectionType && undoData.index !== -1) {
+                    undoData.parentCollection.objects.splice(undoData.index, 0, restoredObject as Expando);
+                  }
+
+                  // Delete nested objects that were hoisted to the parent collection
+                  if (undoData.object instanceof CollectionType) {
+                    undoData.nestedObjects.forEach((obj: Expando) => {
+                      undoData.parentCollection.objects.splice(undoData.parentCollection.objects.indexOf(obj), 1);
+                    });
+                  }
+
+                  // Restore active state if it was active before removal
+                  if (undoData.wasActive) {
+                    await intentPlugin?.provides.intent.dispatch({
+                      action: NavigationAction.ADD_TO_ACTIVE,
+                      data: { id: fullyQualifiedId(restoredObject) },
+                    });
+                  }
+
+                  return { data: true };
+                }
               }
-              break;
+              return { data: false };
```

Comment:
based on the above if statement it seems like the if/elseif conditions cover 100% of cases so this will never be run? could second branch just be an `else`?